src/python/pants/jvm/package/war.py

```diff
@@ -0,0 +1,223 @@
+    BuiltPackageArtifact,
+    OutputPathField,
+    PackageFieldSet,
+)
+from pants.core.target_types import FileSourceField, ResourceSourceField
+from pants.core.util_rules.archive import ZipBinary
+from pants.core.util_rules.source_files import SourceFiles, SourceFilesRequest
+from pants.engine.addresses import Addresses, UnparsedAddressInputs
+from pants.engine.fs import (
+    AddPrefix,
+    CreateDigest,
+    Digest,
+    DigestContents,
+    DigestEntries,
+    FileContent,
+    FileEntry,
+    MergeDigests,
+)
+from pants.engine.internals.native_engine import EMPTY_DIGEST
+from pants.engine.internals.selectors import MultiGet
+from pants.engine.process import BashBinary, Process, ProcessResult
+from pants.engine.rules import Get, collect_rules, rule
+from pants.engine.target import (
+    DependenciesRequest,
+    HydratedSources,
+    HydrateSourcesRequest,
+    SourcesField,
+    Targets,
+    WrappedTarget,
+)
+from pants.engine.unions import UnionRule
+from pants.jvm.classpath import Classpath
+from pants.jvm.target_types import (
+    JvmWarContentField,
+    JvmWarDependenciesField,
+    JvmWarDescriptorAddressField,
+)
+
+logger = logging.getLogger(__name__)
+
+
+@dataclass(frozen=True)
+class PackageWarFileFieldSet(PackageFieldSet):
+    required_fields = (
+        JvmWarDependenciesField,
+        JvmWarDescriptorAddressField,
+    )
+
+    output_path: OutputPathField
+    dependencies: JvmWarDependenciesField
+    descriptor: JvmWarDescriptorAddressField  # TODO: make required
+    content: JvmWarContentField
+
+
+@dataclass(frozen=True)
+class RenderWarDeploymentDescriptorRequest:
+    descriptor: JvmWarDescriptorAddressField
+    owning_address: Address
+
+
+@dataclass(frozen=True)
+class RenderedWarDeploymentDescriptor:
+    digest: Digest
+
+
+@dataclass(frozen=True)
+class RenderWarContentRequest:
+    content: JvmWarContentField
+
+
+@dataclass(frozen=True)
+class RenderedWarContent:
+    digest: Digest
+
+
+@rule
+async def package_war(
+    field_set: PackageWarFileFieldSet,
+    bash: BashBinary,
+    zip: ZipBinary,
+) -> BuiltPackage:
+    classpath = await Get(Classpath, DependenciesRequest(field_set.dependencies))
+    all_jar_files_digest = await Get(Digest, MergeDigests(classpath.digests()))
+
+    prefixed_jars_digest, content, descriptor, make_archive_script_digest = await MultiGet(
+        Get(Digest, AddPrefix(all_jar_files_digest, "__war__/WEB-INF/lib")),
+        Get(RenderedWarContent, RenderWarContentRequest(field_set.content)),
+        Get(
+            RenderedWarDeploymentDescriptor,
+            RenderWarDeploymentDescriptorRequest(field_set.descriptor, field_set.address),
+        ),
+        Get(
+            Digest,
+            CreateDigest(
+                [
+                    FileContent(
+                        "make_war.sh",
+                        textwrap.dedent(
+                            f"""\
+                    mkdir -p __war__/WEB-INF/classes
+                    cd __war__
```

Comment:
Does this require `PATH` to be set?

Reply:
For the `mkdir`? Probably given what we've learned subsequently about OS like nix. Better would be to move the directory creation to the setup of the input digest and avoid needing a `mkdir` call.

Reply:
Yeah looks like just `mkdir`. `cd` is builtin:

```
bash-3.2$ PATH="" cd
bash-3.2$ 
```